File: 2236-maximum-twin-sum-of-a-linked-list/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function pairSum(head: ListNode | null): number {
  let n = 0;
  let now = head;
  while (now) n++, (now = now.next);
  const records: number[] = [];
  let answer = -1;

  let i = 0;
  now = head;
  while (now) {
    if (i < Math.floor(n / 2)) records[i] = now.val;
    else {
      const sum = records[n - 1 - i] + now.val;
      answer = answer > sum ? answer : sum;
    }
    now = now.next;
    i++;
  }
  return answer;
}

File: 2236-maximum-twin-sum-of-a-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        int n = 0;
        ListNode* node = head;
        while(node != NULL) n++, node = node->next;
        node = head;
        vector<int> v(n / 2);
        int ans = 0;
        for(int i = 0; i < n; i++) {
            if(i < n / 2) v[i] = node->val;
            else ans = max(ans, v[n - 1 - i] + node->val);
            node = node->next;
        }
        return ans;
    }
};
